### PR TITLE
Support restore with -XX:CRaCRestoreFrom=PATH

### PR DESCRIPTION
**Pull request by singh264**
Support restore with -XX:CRaCRestoreFrom=PATH that
specifies the path to the checkpoint image.

Issue: https://github.com/eclipse-openj9/openj9/issues/19261
Signed-off-by: Amarpreet Singh <Amarpreet.A.Singh@ibm.com>

### CODE REVIEW

**Comment by singh264** [comment]
@tajila requesting your review. 

It seems like this code change would need to be ported to jdk11, jdk17, jdk22 and jdk-next.

**Comment by tajila** [comment]
@JasonFengJ9 Please review

File: src/java.base/share/native/libjli/java.c [hunk 1]
```diff
@@ -454,6 +455,37 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
...
+    return checkpointDir;
+}
+
+static int restoreFromCheckpoint(char *checkpointDir) {
+    char* argv[7];
+    argv[0] = "criu";
+    argv[1] = "restore";
+    argv[2] = "-D";
+    argv[3] = checkpointDir;
+    argv[4] = "-v4";
```
Comment (tajila) [inline src/java.base/share/native/libjli/java.c:481]: The default logging is v2, but I wouldnt hardcode it. You can detect `-Dopenj9.internal.criu.logLevel` to determine the logging level if its set.

File: src/java.base/share/native/libjli/java.c [hunk 2]
```diff
@@ -454,6 +455,37 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
...
+    argv[5] = "--shell-job";
```
Comment (tajila) [inline src/java.base/share/native/libjli/java.c:482]: you also need to detect `-Dopenj9.internal.criu.unprivilegedMode` to determine if `--unprivileged` needs to be added as well.

File: src/java.base/share/native/libjli/java.c [hunk 3]
```diff
@@ -454,6 +455,37 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argv */
...
+    argv[6] = NULL;
```
Comment (tajila) [inline src/java.base/share/native/libjli/java.c:483]: there is a `--log-file` option as well that you can use `-Dopenj9.internal.criu.logFile` to detect.

**Comment by JasonFengJ9** [comment]
> Normally we do JDKnext first then backport

Please make a PR at https://github.com/ibmruntimes/openj9-openjdk-jdk for reviewing.

In addition, the code added in this change should be decorated with `#if defined(J9VM_OPT_CRAC_SUPPORT)`, and `j9cfg.h` needs to be included.

**Comment by singh264** [comment]
> Please make a PR at https://github.com/ibmruntimes/openj9-openjdk-jdk for reviewing.

I have opened a [PR](https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/780) at https://github.com/ibmruntimes/openj9-openjdk-jdk for reviewing.

**Comment by JasonFengJ9** [comment]
Superseded by https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/199

**Pull request closed without merging**